File: src/lib/strategy/sourceBinding.ts
import { sourceAnalysisDigest } from "../sourceAnalysis";
import { strategySourceEvidenceIds } from "../strategyApproval";
import type { EditorialPlan, EditorialPlanningSnapshot, Job } from "../types";
import { strategySourceBindingSchema, operatorSourceBindingSchema, type PlanningSourceBinding, type StrategySourceBinding } from "./contracts";

type SourceJob = Pick<Job, "id" | "strategyRef" | "sourceAnalysis" | "operatorPlanningContext">;

export function buildStrategySourceBinding(job: SourceJob & { sourceAnalysis: NonNullable<Job["sourceAnalysis"]> }): StrategySourceBinding;
export function buildStrategySourceBinding(job: SourceJob): PlanningSourceBinding;
export function buildStrategySourceBinding(job: SourceJob): PlanningSourceBinding {
  if (!job.sourceAnalysis && job.operatorPlanningContext?.mode === "operator_context") {
    const context = job.operatorPlanningContext;
    if (context.contextDigest !== sourceAnalysisDigest(context.operatorBrief)) throw new Error("operator context digest mismatch");
    return operatorSourceBindingSchema.parse({ ...context, jobId: job.id, strategyRef: job.strategyRef });
  }
  if (!job.strategyRef || !job.sourceAnalysis) throw new Error("pinned strategy and authoritative job analysis required");
  return strategySourceBindingSchema.parse({ jobId: job.id, strategyRef: job.strategyRef, analysisDigest: sourceAnalysisDigest(job.sourceAnalysis), evidenceIds: strategySourceEvidenceIds(job.sourceAnalysis) });
}

export function assertJobSourceBinding(job: SourceJob, snapshot: EditorialPlanningSnapshot, plan?: EditorialPlan): void {
  const expected = buildStrategySourceBinding(job);
  if (!snapshot.sourceBinding || sourceAnalysisDigest(snapshot.sourceBinding) !== sourceAnalysisDigest(expected)) throw new Error("editorial job source binding mismatch");
  const allowed = new Set(expected.evidenceIds);
  if ("mode" in expected) {
    if (plan?.items.some(item => item.evidenceRefs.length)) throw new Error("operator context cannot invent factual evidence");
    return;
  }
  const sourceItems = new Set([...job.sourceAnalysis!.moments, ...job.sourceAnalysis!.angles].map((item) => item.id));
  for (const item of plan?.items ?? []) {
    if (!item.evidenceRefs.length || item.evidenceRefs.some((ref) => !allowed.has(ref)) || !item.evidenceRefs.some((ref) => sourceItems.has(ref))) throw new Error("editorial item evidence is outside authoritative job sources");
  }
}

File: agent/harmonia_agent/source_binding.py
from .canonical import typed_digest


def source_evidence_ids(analysis: dict) -> list[str]:
    return sorted({
        *[item["id"] for item in analysis["moments"]],
        *[ref for item in analysis["moments"] for ref in item["sourceSegmentRefs"]],
        *[item["id"] for item in analysis["angles"]],
        *[ref for item in analysis["angles"] for ref in item["evidenceRefs"]],
    })


def validate_source_binding(binding: dict, job_id: str, strategy_ref: dict, analysis: dict | None, items: list[dict], *, operator_context: dict | None = None) -> None:
    if analysis is None:
        if not operator_context or operator_context.get("mode") != "operator_context":
            raise ValueError("host no-source context authority required")
        if operator_context.get("contextDigest") != typed_digest(operator_context.get("operatorBrief")):
            raise ValueError("operator context digest mismatch")
        if operator_context.get("evidenceIds") != [] or operator_context.get("factualClaimsAllowed") is not False:
            raise ValueError("operator context cannot grant factual evidence")
        if binding != {**operator_context, "jobId": job_id, "strategyRef": strategy_ref}:
            raise ValueError("operator context binding mismatch")
        if any(item.get("evidenceRefs") for item in items):
            raise ValueError("operator context cannot invent factual evidence")
        return
    expected = {"jobId": job_id, "strategyRef": strategy_ref, "analysisDigest": typed_digest(analysis), "evidenceIds": source_evidence_ids(analysis)}
    if binding != expected:
        raise ValueError("editorial job source binding mismatch")
    allowed = set(binding["evidenceIds"])
    source_items = {item["id"] for item in [*analysis["moments"], *analysis["angles"]]}
    for item in items:
        refs = set(item["evidenceRefs"])
        if not refs or not refs <= allowed or not refs & source_items:
            raise ValueError("editorial item evidence is outside authoritative job sources")
